protocol/msgpack/base_iterator.go

```diff
@@ -258,3 +268,19 @@ func (it *baseIterator) checkNumFieldsForTypeWithActual(
 	}
 	return numExpectedFields, numActualFields, true
 }
+
+// bufReader is a buffered reader.
```

Comment:
This is so that the reader we store in `baseIterator` is the actual reader used in the decoder to read the bytes, otherwise if we pass in a non buffered reader, which gets converted to a bufio.Reader inside the decoder, and use the non buffered reader to read the next n bytes when decoding the ids, chances are that we are gonna miss the bytes that are buffered in the internal buffered reader.

It'd be nice if the msgpack decoder exposes the internal reader, or at least the `bufReader` interface, so that for perf-sensitive services the client can decide how to decode the bytes (in place or making a new copy of a byte slice). But it's not, so for now this is the workaround.